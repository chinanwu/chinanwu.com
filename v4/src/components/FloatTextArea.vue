<template>
  <div class="FloatTextArea">
    <label
      class="FloatTextArea__label"
      :class="{ 'FloatTextArea__label--float': float }"
      :id="'floatInputLabel-' + label"
      :for="'floatInput-' + label"
      :data-float="float"
    >
      {{ label }}
    </label>
    <textarea
      class="FloatTextArea__textarea"
      :id="'floatInput-' + label"
      :aria-labelledby="'floatInputLabel-' + label"
      v-model="inputValue"
      @focusin="handleFocusIn"
      @focusout="handleFocusOut"
    />
  </div>
</template>

<script>
export default {
  name: "FloatTextArea",
  props: {
    label: {
      type: String,
      required: true
    }
  },
  data() {
    return {
      float: false,
      inputValue: ""
    };
  },
  methods: {
    handleFocusIn() {
      this.float = true;
    },
    handleFocusOut() {
      if (this.inputValue === "") {
        this.float = false;
      }
    }
  }
};
</script>

<style lang="less" scoped>
@import "../styling/colours.less";

.FloatTextArea {
  position: relative;
  margin-top: 3rem;
}

.FloatTextArea__label {
  position: absolute;
  font-size: 2.4rem;
  transition: all 0.2s;
  padding-left: 1rem;

  &[data-float="true"] {
    transform: translateY(-2rem);
    font-size: 1.6rem;
    padding-left: 0;
  }
}

.FloatTextArea__textarea {
  font-size: 2.4rem;
  width: 100%;
}
</style>
