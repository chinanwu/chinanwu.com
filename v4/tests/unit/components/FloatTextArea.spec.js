import { shallowMount } from "@vue/test-utils";
import FloatTextArea from "@/components/FloatTextArea.vue";

describe("FloatTextArea", () => {
  describe("rendering", () => {
    it("renders", () => {
      const wrapper = shallowMount(FloatTextArea, {
        props: {
          label: "test"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe("behaviour", () => {
    it("floats if focused in on", () => {
      const wrapper = shallowMount(FloatTextArea, {
        props: {
          label: "label"
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      expect(wrapper.vm.float).toBeFalsy();
      wrapper.find("#floatTextArea-label").trigger("focusin");
      expect(wrapper.vm.float).toBeTruthy();
    });

    it("stops floating if focus moves away with no input", () => {
      const wrapper = shallowMount(FloatTextArea, {
        props: {
          label: "label"
        },
        data() {
          return {
            float: true
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#floatTextArea-label").trigger("focusout");
      expect(wrapper.vm.float).toBeFalsy();
    });

    it("keep floating if focus moves away with input", () => {
      const wrapper = shallowMount(FloatTextArea, {
        props: {
          label: "label"
        },
        data() {
          return {
            float: true,
            inputValue: "something"
          };
        }
      });
      expect(wrapper.element).toMatchSnapshot();
      wrapper.find("#floatTextArea-label").trigger("focusout");
      expect(wrapper.vm.float).toBeTruthy();
      expect(wrapper.emitted().change).toEqual([["something"]]);
    });
  });
});
