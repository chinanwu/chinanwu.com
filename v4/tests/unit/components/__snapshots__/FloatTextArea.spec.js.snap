// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FloatTextArea behaviour floats if focused in on 1`] = `
<div
  class="FloatTextArea"
>
  <label
    class="FloatTextArea__label"
    data-float="false"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <textarea
    aria-labelledby="floatTextAreaLabel-label"
    class="FloatTextArea__textarea"
    id="floatTextArea-label"
  />
</div>
`;

exports[`FloatTextArea behaviour keep floating if focus moves away with input 1`] = `
<div
  class="FloatTextArea"
>
  <label
    class="FloatTextArea__label FloatTextArea__label--float"
    data-float="true"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <textarea
    aria-labelledby="floatTextAreaLabel-label"
    class="FloatTextArea__textarea"
    id="floatTextArea-label"
  />
</div>
`;

exports[`FloatTextArea behaviour stops floating if focus moves away with no input 1`] = `
<div
  class="FloatTextArea"
>
  <label
    class="FloatTextArea__label FloatTextArea__label--float"
    data-float="true"
    for="floatInput-label"
    id="floatInputLabel-label"
  >
    label
  </label>
  <textarea
    aria-labelledby="floatTextAreaLabel-label"
    class="FloatTextArea__textarea"
    id="floatTextArea-label"
  />
</div>
`;

exports[`FloatTextArea rendering renders 1`] = `
<div
  class="FloatTextArea"
>
  <label
    class="FloatTextArea__label"
    data-float="false"
    for="floatInput-test"
    id="floatInputLabel-test"
  >
    test
  </label>
  <textarea
    aria-labelledby="floatTextAreaLabel-test"
    class="FloatTextArea__textarea"
    id="floatTextArea-test"
  />
</div>
`;
